streaming/base/dataset.py

```diff
@@ -747,138 +933,172 @@ def get_item(self, sample_id: int) -> Any:
             # because there can only be one epoch at once.
-            if state.is_stopped:
+            if it.is_exiting():
                 break
 
             # If we're out of samples this epoch, exit this thread because we are done downloading.
-            if state.ready_index == state.total:
+            if it.ready_index == it.total:
                 break
 
             # If we are requested to only pre-download so many samples, if we have as many or more
             # downloaded already, we wait and check again later.
             if self.predownload is not None:
-                samples_ahead = state.ready_index - state.yield_index
+                samples_ahead = it.ready_index - it.yield_index
                 if self.predownload <= samples_ahead:
                     sleep(TICK)
                     continue
 
             # If we hit -1, we skip.
-            sample_id = state.sample_ids[state.ready_index]
+            sample_id = it.sample_ids[it.ready_index]
             if sample_id == -1:
-                state.ready_index += 1
+                it.ready_index += 1
                 continue
 
-            # Download and decompress the shard for this sample, if not already done.
+            # Wait for the shard for this sample to be downloaded and decompressed, if not already.
             shard_id, _ = self.spanner[sample_id]
-            while shard_states[shard_id] != _ShardState.DOWNLOADED:
-                sleep(TICK)
-            state.ready_index += 1
+            self.download_shard(shard_id, True)
+
+            # Step forward one sample.
+            it.ready_index += 1
+
+        # Note that we exited.
+        it.on_exit()
+
+    def _each_sample_id(self, it: _IterState) -> Iterator[int]:
+        """Iterate over our samples while waiting for them to download first.
+
+        This method is entered at the beginning of each epoch, and exits either when out of samples
+        or when a new epoch is started, calling exit() on its state (only one epoch is valid at a
+        time).
 
-    def _each_sample(self, sample_ids: NDArray[np.int64]) -> Iterator[int]:
-        """Iterate over each sample ID, while downloading ahead in the background.
+        Each worker has its own yield loop, which iterates behind the download and ready threads.
 
         Args:
-            sample_ids (NDArray[np.int64]): The sample IDs to download and iterate.
+            it (_IterState): State of __iter__.
 
         Returns:
-            Iterator[int]: Each sample ID, having been downloaded.
+            Iterator[int]: Each sample, having been downloaded.
         """
-        self._partition_state = _PartitionState(sample_ids)
-        Thread(target=self._download_thread, args=(self._partition_state,), daemon=True).start()
-        Thread(target=self._ready_thread, args=(self._partition_state,), daemon=True).start()
-        yield from self._partition_state
+        # Yield loop.
+        while True:
```

Comment:
I get scared when I see infinite loop. Can we have some time to track and kill if it expires ?

Reply:
This loop is the epoch, no way to limit this